### PR TITLE
Collator set refactor

### PR DESCRIPTION
**Pull request by vieira-giulia**
Functions added:

- [x]  add_invulnerable

- [x]  remove_invulnerable

- [x]  sort invulnerables in set_invulnerables

- [x] Tests

- [x] Benchmarking 

Weights are still arbitrary values

### CODE REVIEW

**Comment by vieira-giulia** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2316329

**Comment by gilescope** [comment]
if we have add and remove invulnerables then we potentially could remove set_invulnerables. That said I think I am inclined to keep it.

**Comment by bkchr** [comment]
@vieira-giulia can you please explain why we need this? `set_invulnerables` already can be used to emulate add/remove.

**Comment by joepetrowski** [comment]
> @vieira-giulia can you please explain why we need this? `set_invulnerables` already can be used to emulate add/remove.

Yes, `set` can emulate add/remove, but the idea is that we want to use add/remove more regularly as we launch more system chains because it doesn't make sense for collators to place deposits on multiple chains for which they collate.

So far, `Invulnerables` has been used to set the genesis collators, but we expect most paras to have 15-20 Invulnerables and then an election for some extra slots, where the Invulnerables would be chosen based on reputation for having collated other chains or been a validator. So we expect governance proposals, like "add <my-account> as an invulnerable on some para", that may coincide with other proposals to remove an invulnerable.

File: pallets/collator-selection/src/lib.rs [hunk 1]
```diff
@@ -306,13 +312,91 @@ pub mod pallet {
...
+		 		);
+
+			// remove invulnerable from invulnerables list
+			<Invulnerables<T>>::try_mutate(|invulnerables| -> DispatchResult {
+				let pos = invulnerables
+					.binary_search(&to_remove)
+					.ok()
+					.ok_or(Error::<T>::NotInvulnerable)?;
+				invulnerables.remove(pos);
+				invulnerables.sort();
```
Comment (joepetrowski) [inline pallets/collator-selection/src/lib.rs:389]: You shouldn't need to sort on remove. If you remove an element from a sorted list it is still sorted.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2371047

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2371242

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2371620

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2371790

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2490400

**Comment by vieira-giulia** [comment]
bot benchmark

**Comment by vieira-giulia** [comment]
bot bench-all

**Comment by vieira-giulia** [comment]
bot bench $ pallet dev pallet_collator_selection

**Comment by vieira-giulia** [comment]
bot bench $ pallet dev pallet_collator_selection

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2495863

File: pallets/collator-selection/src/lib.rs [hunk 2]
```diff
@@ -306,13 +312,77 @@ pub mod pallet {
...
+			// ensure this new invulnerable has registred a validator key
+			let validator_key = T::ValidatorIdOf::convert(new.clone())
+				.ok_or(Error::<T>::NoAssociatedValidatorId)?;
+			ensure!(
+				T::ValidatorRegistration::is_registered(&validator_key),
+				Error::<T>::ValidatorNotRegistered
+			);
+
+			<Invulnerables<T>>::try_mutate(|invulnerables| -> DispatchResult {
+				invulnerables.try_push(new.clone()).map_err(|_| Error::<T>::Unknown)?;
```
Comment (xlc) [inline pallets/collator-selection/src/lib.rs:350]:
```suggestion
				invulnerables.try_push(new.clone()).map_err(|_| Error::<T>::TooManyInvulnerables)?;
```

**Comment by vieira-giulia** [comment]
bot bench $ pallet statemint-dev pallet_collator_selection

**Comment by vieira-giulia** [comment]
bot bench $ pallet statemint asset pallet_collator_selection

**Comment by vieira-giulia** [comment]
bot bench $ pallet statemint asset pallet_collator_selection

**Comment by vieira-giulia** [comment]
bot bench $ pallet statemint asset pallet_collator_selection

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2669528

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2669612

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2692978

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2693157

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2693210

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2757264

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2757349

**Comment by joepetrowski** [comment]
Will update

**Pull request closed without merging**